runtime/src/commonMain/kotlin/pbandk/internal/json/JsonValueDecoder.kt

```diff
@@ -74,7 +81,26 @@ internal class JsonValueDecoder(private val jsonConfig: JsonConfig) {
                 "Integers with leading zeros are not allowed"
             )
         }
-        body(content)
+
+        if (content.last() == ' ') {
+            throw NumberFormatException(
+                "Integers must not have trailing space"
+            )
+        }
+
+        val numberTrailingZeroes = """\.0+$""".toRegex()
+        val numberScientificNotation = """-?(?:\d+)(\.\d+)?[eE](\d+)$""".toRegex()
+
+        var contentWithoutTrailingZero = content.replace(numberTrailingZeroes, "")
+        numberScientificNotation.find(contentWithoutTrailingZero)?.let {
+            val mantissaDigits = it.groupValues[1].length - 1
+            val decade = it.groupValues[2].toInt()
+            if (decade >= mantissaDigits) {
+                contentWithoutTrailingZero = contentWithoutTrailingZero.toDouble().toLong().toString()
```

Comment:
I have a couple concerns with this approach:

1. I think this code doesn't handle all valid values correctly. For example, `3.10e1`. I'm sure it can be made to work with some additional checks, but that brings me to my second concern...

2. This might be premature optimization, but I'm worried about the performance impact of doing regex matching for parsing every number.

Do you think there's a way to handle this without regexs? or maybe a way to have a "fast path" for regular integers and fall back to the "slow path" with regexs only if the number is using floating-point or exponential notation?

Reply:
[1] is a valid concern. Regex matching is to be avoided if possible since it leads to weird edge cases like you found yourself. [2] is not really an issue if we compile the regexes on object level, they can be pretty fast.

As an initial approach, I looked into using `toDouble()` however as suspected `toDouble()` is very lenient on input, effectively making it useless. But I have another idea, will try it out. Perhaps we can do away with regex.

I like the idea of having fast path for regular integers 👍 either way

Reply:
Had a nice solution with `BigDecimal`:

```
val bigDecimal = content.toBigDecimal()
if (bigDecimal.isWhole()) {
      body(bigDecimal.toPlainString())
}
```

However there's no multiplatform support for it :( And in the absence of an unboxing type that can safely unbox 65+ bits, `toDouble` coerces the value to `toXXX()` that always fits the range.

Reply:
It's a shame that `toDouble()` is so lenient. If we can have a regex-free fast path for integers and pre-compile the regexs for the slow path, then I think it's fine to keep the regexs for the slow path.

Reply:
Have re-added the regex, once #114 is completed we can decided if we go with https://github.com/ionspin/kotlin-multiplatform-bignum 
It's still pretty fresh with some caveats, but with `toDouble()` to test for `NumberFormat` (the library does not handle invalid formats) it can be a nice alternative to regex, as per the code above.